Add azure servicebus notify service

hfurubotten opened the pull request:
## Description:

Adds an integration to Azure Service Bus. Will allow users to send custom messages from Home Assistant for asynchronously processing in Azure or any platform that can read from Azure Service Bus afterwards. 

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10756

## Example entry for `configuration.yaml` (if applicable):
```yaml
# Example configuration.yaml entry
notify:
  - platform: azure_service_bus
    name: Test Topic
    connection_string: !secret servicebus_connection_string
    topic: t-test
  - platform: azure_service_bus
    name: Test Queue
    connection_string: !secret servicebus_connection_string
    queue: q-test
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

hfurubotten committed:
Add azure servicebus notify service

hfurubotten committed:
files added to .coveragerc

hfurubotten committed:
fix: import content type from const

hfurubotten committed:
Moved imports to top level

MartinHjelmare commented on homeassistant/components/azure_servicebus/manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "domain": "azure_servicebus",
```
Should we name it `azure_service_bus`, as it's three words?

hfurubotten replied:
Tried to research what the correct naming is, with or without separation in `service bus`, but it seems even microsoft isn't consistent on this. Naming in text seems to be two words, but when they use it in libraries and code patterns they use one word. 
If three words in the integration name is more suited for HA, I will of course change it!

MartinHjelmare replied:
On Azure website they write the name in three words.
https://docs.microsoft.com/en-us/azure/service-bus-messaging/

And looking at eg `ServiceBusClient` class it looks like the intent is to separate service and bus since Bus is capitalized in the PascalCased name.

I suggest we go with `azure_service_bus` as integration name.

hfurubotten replied:
👍 Will update both PR's tomorrow

MartinHjelmare commented on homeassistant/components/azure_servicebus/manifest.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "domain": "azure_servicebus",
+  "name": "Azure Service Bus",
+  "documentation": "https://www.home-assistant.io/components/azure_servicebus",
```
Please replace `components` with `integrations`.

MartinHjelmare commented on homeassistant/components/azure_servicebus/manifest.json:
```diff
@@ -0,0 +1,10 @@
+    "azure-servicebus==0.50.1"
+  ],
+  "dependencies": [],
+  "codeowners": []
```
Please add yourself as codeowner.

MartinHjelmare commented on homeassistant/components/azure_servicebus/notify.py:
```diff
@@ -0,0 +1,83 @@
+        client = servicebus.get_topic(topic_name)
+    else:
+        _LOGGER.error(
+            "Could not find a topic or queue from the configuration. Please provide one!"
```
Please add `cv.has_at_least_one_key` in the config schema instead.

hfurubotten replied:
Didn't know about that one, thanks for enlightening me! Have updated the configuration 😃

MartinHjelmare commented on homeassistant/components/azure_servicebus/notify.py:
```diff
@@ -0,0 +1,83 @@
+    topic_name = config.get(CONF_TOPIC_NAME)
+
+    servicebus = ServiceBusClient.from_connection_string(
+        connection_string, loop=hass.loop
```
Does the library support asyncio? Should we use the home assistant async api instead, ie `async_get_service`?

hfurubotten replied:
The library does support asyncio, but it doesn't do any IO or async calls during setup so I picked the sync method because of that. Changed it to `async_get_service` now and tested it, and it works. I'm not sure I have completely understood everything around asyncio, so please correct me if that doesn't work.

MartinHjelmare replied:
It's good to stay in async context as long as we can. There's some overhead to switching context.

If the library supports asyncio it's most likely intended to run within the event loop by default, even when it's not strictly necessary.

Many asyncio functions are not threadsafe and must be run within the event loop to work. Eg getting the current event loop with `asyncio.get_event_loop` works inside the event loop, in callbacks and coroutines.

MartinHjelmare replied:
We need to change back to use `get_service` since the library is implemented very weirdly. See below.

hfurubotten replied:
Changed back to sync method now, and hopefully added a good enough comment to warn about this.

MartinHjelmare commented on homeassistant/components/azure_servicebus/notify.py:
```diff
@@ -0,0 +1,83 @@
+        dto = {ATTR_MESSAGE: message}
+
+        if ATTR_TITLE in kwargs:
+            dto[ATTR_TITLE] = kwargs[ATTR_TITLE]
```
Are the keys in the dto defined by the Azure api, or are they arbitrary? If they are Azure api specific we should not use the same constants as in the home assistant api. That could break the integration if we change the home assistant constants.

hfurubotten replied:
Every field into the dto is arbitrary, so it shouldn't make a difference if the names changes, but you have a good point that if the constants change it would also make the output change. 
Have added separate constants for the output to servicebus, so if the internals change it doesn't break anything anyone has built on top of this.

hfurubotten committed:
Code review fixes
+ added code owner
+ fixed config validation with has at least one
+ seperate attributes for dto to asb
* fixed doc link
* async setup instead of sync

MartinHjelmare commented on homeassistant/components/azure_servicebus/notify.py:
```diff
@@ -0,0 +1,83 @@
+    )
+
+    if queue_name:
+        client = servicebus.get_queue(queue_name)
```
I'm looking at the library source. It's very weird. Getting the queue seems to be using the requests library, ie sync I/O.

https://github.com/Azure/azure-sdk-for-python/blob/c648cc4d18d2052bb57a6d8259f56ffa40c94b39/sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_client.py#L117-L119

While at the same time they are using asyncio functions that must be run within the event loop when creating the client:
https://github.com/Azure/azure-sdk-for-python/blob/c648cc4d18d2052bb57a6d8259f56ffa40c94b39/sdk/servicebus/azure-servicebus/azure/servicebus/aio/async_client.py#L58

Since we're passing the loop explicitly it should be safe to run this in sync context to avoid making sync I/O inside the event loop.

I'm not sure what they are thinking when creating the library.

hfurubotten replied:
They upgraded the library with breaking changes a little while back, and it seems like they try to combine 3 different ways of using the clients to not break too much for anyone.
Looks like they perhaps wasn't careful enough while reusing some parts...

MartinHjelmare commented on homeassistant/components/azure_servicebus/notify.py:
```diff
@@ -0,0 +1,83 @@
+        client = servicebus.get_queue(queue_name)
+    elif topic_name:
+        client = servicebus.get_topic(topic_name)
+    else:
```
This case will never happen.

hfurubotten committed:
rename all the things
- removed too many ifs
* changed setup back to sync
+ comment about sync IO in lib

MartinHjelmare reviewed (approved):
Looks great!

MartinHjelmare commented on homeassistant/components/azure_service_bus/notify.py:
```diff
@@ -0,0 +1,83 @@
+    )
+
+    if queue_name:
+        client = servicebus.get_queue(queue_name)
```
We should except the possible exceptions that this can raise, and take appropriate action. Eg if the resource is not found is unrecoverable, right? Connection error could perhaps be recoverable. Can we rule out that the connection error is due to incorrect connection_string?

`ServiceBusConnectionError`
`ServiceBusResourceNotFound`

hfurubotten replied:
I think both of them is will be raised if the user sends in incorrect information on the connection string. The `.from_connection_string(...)` method above til validate the format and that all the needed parts are included, but won't check if the resources are there before getting the client. For instance if the domain is incorrect it would be raised as `ServiceBusConnectionError`, and would need the user to update the configuration or fix the resource in Azure and give new config. 

Even if I manage to research which inner exceptions would be retryable due to intermittent connection errors, it doesn't seem like raising `PlatformNotReady` work during setup on notification services? I think it might be best to just log the error better than what the general exception handling in notify component does, to inform the user as best as possible, and return `None` to stop the setup. Agree or am I missing something?

MartinHjelmare replied:
Sounds good!

MartinHjelmare commented on homeassistant/components/azure_service_bus/notify.py:
```diff
@@ -0,0 +1,83 @@
+    if queue_name:
+        client = servicebus.get_queue(queue_name)
+    else:
+        client = servicebus.get_topic(topic_name)
```
As above.

MartinHjelmare reviewed (changes requested):
Sorry, I found a couple of more comments.

MartinHjelmare commented on homeassistant/components/azure_service_bus/notify.py:
```diff
@@ -0,0 +1,83 @@
+
+        queue_message = Message(json.dumps(dto))
+        queue_message.properties.content_type = CONTENT_TYPE_JSON
+        await self._client.send(queue_message)
```
This can raise `MessageSendFailed`. We should catch that and log error.

hfurubotten committed:
More informative logging

MartinHjelmare commented on homeassistant/components/azure_service_bus/notify.py:
```diff
@@ -0,0 +1,104 @@
+        else:
+            client = servicebus.get_topic(topic_name)
+    except (ServiceBusConnectionError, ServiceBusResourceNotFound) as err:
+        _LOGGER.exception(
```
We usually limit to error level to avoid the full stack trace print.

hfurubotten committed:
logging exception -> error

MartinHjelmare reviewed (approved):
Good!
